### PR TITLE
#92: first draft Portrait image store

### LINKED ISSUE #92: Store identity portraits
Note: this issue is split from #70 

When login in for the first time, a portrait image might be provided in the identity data. In this case we need to download and store this image. This originally was a part of the issue mentioned above, but was split of to make make its scope smaller.

### PR DESCRIPTION
**Pull request by JohnatMasking**
Fixes #92 

Changes proposed in this pull request:
- download picture from Identity
- store image in filestorage
- added tests
-

@MaskingTechnology/comify

### CODE REVIEW

File: src/domain/image/data/retrieve.ts [hunk 1]
```diff
@@ -1,7 +1,12 @@
 
-import ImageData from './ImageData';
+import database from '../../../integrations/database/module';
+import type ImageData from './ImageData';
+import { RECORD_TYPE } from './constants.js';
+import createData from './createData';
 
-export default async function retrieve(id: string): Promise<ImageData>
+export default async function retrieve(imageId: string): Promise<ImageData>
```
Comment (BasMasking) [inline src/domain/image/data/retrieve.ts:7]: The variable `id` is used in all cases for retrieve.

File: src/domain/image/download.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
+
+import ImageData from './data/ImageData';
+import { UnsupportedContentSize, UnsupportedMimeType } from './errors';
+import save from './save';
+
+async function checkImage(picture: string): Promise<void>
+{
+    const response = await fetch(picture, { method: 'HEAD' });
+    const mimeType = response.headers.get('content-type');
```
Comment (BasMasking) [inline src/domain/image/download.ts:9]: The `content-type` should be a definition

File: src/domain/image/download.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    if (mimeType === null || ['image/gif', 'image/jpeg', 'image/png'].includes(mimeType) === false)
```
Comment (BasMasking) [inline src/domain/image/download.ts:11]: The supported mime types should be a definition

File: src/domain/image/download.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        throw new UnsupportedMimeType;
+    }
+
+    const length = response.headers.get('content-length');
```
Comment (BasMasking) [inline src/domain/image/download.ts:16]: The `content-length` should be a definition

File: src/domain/image/download.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    const size = Number(length);
+
+    if (length === null || size > 1024 * 1024 * 5)
```
Comment (BasMasking) [inline src/domain/image/download.ts:19]: The maximum size should be a definition

File: src/domain/image/errors.ts
Comment (BasMasking) [inline src/domain/image/errors.ts:1]: The errors in the domain are separated in an error folder (new insights). There should be an `error` folder with the `UnsupportedFileType` and `UnsupportedFileSize`.

File: src/domain/image/download.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    {
+        throw new UnsupportedContentSize;
+    }
+}
+
+export default async function download(picture: string, imageType: string): Promise<ImageData>
```
Comment (BasMasking) [inline src/domain/image/download.ts:25]: The implementation of this function is not finished yet, but I've made a rough draft for you.

File: src/domain/image/files/store.ts [hunk 7]
```diff
@@ -0,0 +1,7 @@
+
+import fileStorage from '../../../integrations/filestorage/module';
+
+export default async function store(path: string, buffer: Buffer): Promise<void>
+{
+    await fileStorage.writeFile(path, buffer);
```
Comment (BasMasking) [inline src/domain/image/files/store.ts:6]: We should check if the file exists, before overwriting it.

**Review by BasMasking — DISMISSED** [review]

File: test/domain/authentication/_fixtures/login.fixture.ts [hunk 8]
```diff
@@ -88,8 +91,42 @@ const LOGINS: Record<string, Identity> =
...
+        email_verified: false
+    },
+    NAME_WITH_TOO_MANY_SIMILAR_NICKNAME: {
+        name: creator4.fullName as string,
+        nickname: 'TooManyNickName',
+        picture: undefined,
+        email: 'toomanynickname@mail.com',
+        email_verified: false
+    },
+    NAME_WITH_A_VALID_PICTURE_URL: {
```
[the remark this answers is not in the document]
Comment (JohnatMasking) [reply test/domain/authentication/_fixtures/login.fixture.ts:109]: Changed to id.

File: src/domain/image/data/retrieve.ts (hunk 1 above)
Comment (JohnatMasking) [reply src/domain/image/data/retrieve.ts:7]: Changed to id.

File: src/domain/image/download.ts (hunk 2 above)
Comment (JohnatMasking) [reply src/domain/image/download.ts:10]: changed to a constant

File: src/domain/image/download.ts (hunk 3 above)
Comment (JohnatMasking) [reply src/domain/image/download.ts:12]: changed to a constant

File: src/domain/image/download.ts (hunk 4 above)
Comment (JohnatMasking) [reply src/domain/image/download.ts:17]: changed to a constant

File: src/domain/image/download.ts (hunk 5 above)
Comment (JohnatMasking) [reply src/domain/image/download.ts:20]: changed to a constant

File: src/domain/image/download.ts (hunk 6 above)
Comment (JohnatMasking) [reply src/domain/image/download.ts:26]: Finished function

File: src/domain/image/files/store.ts (hunk 7 above)
Comment (petermasking) [reply src/domain/image/files/store.ts:6]: This is fixed in the `save` function.

**Review by petermasking — APPROVED** [review]

**Pull request merged**